lib/client/db/postgres/mcp/mcp.go

```diff
@@ -0,0 +1,253 @@
+}
+
+// RunQuery tool function used to execute queries on databases.
+func (s *Server) RunQuery(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	uri := request.Params.Arguments[queryToolDatabaseParam].(string)
+	sql := request.Params.Arguments[queryToolQueryParam].(string)
+
+	db, err := s.getDatabase(uri)
+	if err != nil {
+		return s.wrapErrorResult(ctx, nil, err)
+	}
+
+	// TODO(gabrielcorado): ensure the connection used is consistent for the
+	// session, making most of its queries to be present in a single audit
+	// session/recording.
+	rows, err := db.pool.Query(ctx, sql)
+	if err != nil {
+		return s.wrapErrorResult(ctx, db.ExternalErrorRetriever, err)
+	}
+
+	// Returned rows are being closed by this function.
+	result, err := buildQueryResult(rows)
+	if err != nil {
+		return s.wrapErrorResult(ctx, db.ExternalErrorRetriever, err)
+	}
+
+	return mcp.NewToolResultText(result), nil
+}
+
+func (s *Server) wrapErrorResult(ctx context.Context, externalRetriever dbmcp.ExternalErrorRetriever, toolErr error) (*mcp.CallToolResult, error) {
+	s.logger.ErrorContext(ctx, "error while querying database", "error", toolErr)
+	out, err := json.Marshal(RunQueryResult{ErrorMessage: dbmcp.FormatErrorMessage(externalRetriever, toolErr).Error()})
+	return mcp.NewToolResultError(string(out)), trace.Wrap(err)
+}
+
+// buildQueryResult takes a the response from pgx and converts into a JSON
+// format (which will be returned to LLMs).
+func buildQueryResult(rows pgx.Rows) (string, error) {
+	// Just ensure the rows is always closed. It is safe if this is called
+	// multiple times.
+	defer rows.Close()
+
+	var data []map[string]any
+	columns := rows.FieldDescriptions()
+
+	for rows.Next() {
+		values, err := rows.Values()
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+
+		item := make(map[string]any, len(values))
+		for i, v := range values {
+			item[columns[i].Name] = v
+		}
+
+		data = append(data, item)
+	}
+
+	// Close the rows to finish consuming it. Depending on the its type
+	// we can only collect the command tag after rows is closed.
+	rows.Close()
+	commandTag := rows.CommandTag()
+
+	// Initialize the slice so the resulting JSON will have an empty array
+	// instead of null.
+	if len(data) == 0 && commandTag.Select() {
+		data = []map[string]any{}
+	}
+
+	out, err := json.Marshal(RunQueryResult{
+		Data:      data,
+		RowsCount: int(commandTag.RowsAffected()),
+	})
+	return string(out), trace.Wrap(err)
+}
+
+func (s *Server) getDatabase(uri string) (*database, error) {
+	// The caller might refer to the database by its name instead of its URI.
+	// For this case we fallback into generating the URI as the databases are
+	// organized by URI.
+	if !dbmcp.IsDatabaseResourceURI(uri) {
+		uri = dbmcp.DatabaseResourceURI(uri)
+	}
+
+	db, ok := s.databases[uri]
+	if !ok {
+		return nil, trace.NotFound("Database %q not found. Only Teleport databases resources can be used for queries.", uri)
```

Comment:
Should we show available ones in this message?

Reply:
I don't think that would be necessary. Reinforcing it it has to be a database resource URI is more reliable as models can request users to attach resources or directly call the list databases tool.